### PR TITLE
common: extract options into yaml

### PR DESCRIPTION
**Pull request by tchaikov**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-04-11_12:36:03-rados-wip-kefu-testing-2021-04-11-1826-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-04-12_02:29:32-rados-wip-kefu-testing-2021-04-12-0057-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-04-12_08:01:43-rados-wip-kefu-testing-2021-04-12-1256-distro-basic-smithi/

multi-site failure tracked by https://tracker.ceph.com/issues/50218

**Comment by tchaikov** [comment]
@liewegas @jdurgin @neha-ojha @cbodley @batrick @trociny ping?

**Comment by liewegas** [comment]
This looks great!  Is the legacy_config_opts.h generation going to be incorporated too?

**Review by batrick — APPROVED** [review]
Great change!

File: src/common/options/build_options.cc [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    for (auto &o : options) {
+      o.add_service(svc);
+      result.push_back(std::move(o));
+    }
+  };
+
+  ingest(get_rgw_options(), "rgw");
+  ingest(get_rbd_options(), "rbd");
+  ingest(get_rbd_mirror_options(), "rbd-mirror");
+  ingest(get_immutable_object_cache_options(), "immutable-objet-cache");
```
Comment (jdurgin) [inline src/common/options/build_options.cc:29]:
```suggestion
  ingest(get_immutable_object_cache_options(), "immutable-object-cache");
```

File: src/common/options/immutable-object-cache.yaml.in [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+---
+
+options:
+- name: immutable_object_cache_path
+  type: str
+  level: advanced
+  desc: immutable object cache data dir
+  default: /tmp/ceph_immutable_object_cache
+  services:
+  - immutable-objet-cache
```
Comment (jdurgin) [inline src/common/options/immutable-object-cache.yaml.in:11]: s/objet/object/g in this file

File: src/common/options/build_options.cc (hunk 1 above)
Comment (tchaikov) [reply src/common/options/build_options.cc]: @jdurgin thanks for your review. i created a separate commit for fixing this typo. as it exists in master as well.

**Comment by tchaikov** [comment]
changelog:

- rebased against master
- add a commit to s/immutable-objet-cache/immutable-object-cache/

**Comment by jdurgin** [comment]
Looks great! could you add tests for y2c.py as well? there are many fields that we wouldn't notice were missing if the build succeeded, and I don't think this is the last time this script will change

**Comment by jdurgin** [comment]
given the rebasing difficulty I'm fine with tests being a follow-up pr

**Review by jdurgin — APPROVED** [review]

File: src/common/options/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  list(APPEND common_options_srcs
+    ${src_file})
+  set(common_options_srcs ${common_options_srcs} PARENT_SCOPE)
+endfunction()
+
+set(osd_erasure_code_plugins "jerasure" "lrc")
+if(WITH_EC_ISA_PLUGIN)
+  list(APPEND osd_erasure_code_plugins "isa")
+endif()
+string(REPLACE ";" " " osd_erasure_code_plugins "${osd_erasure_code_plugins}")
```
Comment (cbodley) [inline src/common/options/CMakeLists.txt:45]: where does this cmake variable get applied to the configuration?

this looks related to a new vstart.sh issue where the mon fails to start with:
> load: jerasure load: lrc load dlopen(build/lib/libec_isa.so): build/lib/libec_isa.so: cannot open shared object file: No such file or directory

ceph-mon seems to be taking `osd_erasure_code_plugins` directly from the yaml (`default: jerasure lrc isa`) and tries to load `isa` even though `WITH_EC_ISA_PLUGIN` is disabled

Comment (tchaikov) [reply src/common/options/CMakeLists.txt:45]: @cbodley thank you for pointing out this regression. should be fixed by #40971 .

**Pull request merged**